service/migration_manager.cc

```
@@ -1080,20 +1080,19 @@ static future<schema_ptr> get_schema_definition(table_schema_version v, locator:
             // referenced by the incoming request.
             // That means the column mapping for the schema should always be inserted
             // with TTL (refresh TTL in case column mapping already existed prior to that).
-            auto us = s.unfreeze(db::schema_ctxt(proxy));
+            // We don't set the CDC schema here because it's not included in the RPC, but with
+            // schema on raft we shouldn't get here anyway because we do a group0 barrier which
+            // should ensure the schema is added to the registry together with the CDC schema.
+            auto us = s.unfreeze(db::schema_ctxt(proxy), nullptr);
```

I'm not sure about this
are schema pulls always disabled now? or is it possible we pull some schema, maybe during upgrade or recovery, and use it to generate cdc?

---

Schema pulls are always disabled if a cluster is in raft mode. Clusters that are not in raft mode will have schema pulls enabled. During upgrade schema pulls will be enabled up until the point the cluster is fully upgraded. New clusters on all supported versions will have it always disabled. Hope this covers all scenarios.

---

thanks. so if I understand correctly we still need to support the cluster running not in raft mode.
in this case I think we may not always be able to set a compatible cdc schema, and if it's not set we can revert to using the current cdc schema like before.

---

I updated the PR to handle the case that we run without raft and get a base schema without its cdc schema by schema pull.
in this case the pointer to the cdc schema can be null, and the cdc transformer will lookup the current cdc schema from the db.

---

We do not support schema changing while in recovery.

---

Probably the test should check if schema version actually changed. I think it will not. You are right that we need to throw an error in this case though.